endaq/calc/shock.py

```diff
@@ -62,75 +219,170 @@ def rel_displ(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFr
         - `SciPy transfer functions <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.TransferFunction.html>`_
           Documentation for the transfer function class used to characterize the
           relative displacement calculation.
-        - `SciPy biquad filter <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html>`_
+        - `SciPy biquad filter <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html>`__
           Documentation for the biquad function used to implement the transfer
           function.
+        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`
+          Explicit implementations of digital filter coefficients for shock spectra.
     """
-    dt = utils.sample_spacing(accel)
-    tf = _rel_displ_transfer_func(omega, damp, dt)
+    T = utils.sample_spacing(accel)
+    Q = 1./(2.*damp)
 
     return accel.apply(
-        functools.partial(scipy.signal.lfilter, tf.num, tf.den, axis=0),
+        functools.partial(
+                scipy.signal.lfilter,
+                *_rel_displ_coeffs(omega, Q, T),
+                axis=0,
+                ),
         raw=True,
-    )
+        )
 
 
-def _abs_accel_transfer_func(
-    omega: float, damp: float = 0.0, dt: float = 1.0
-) -> scipy.signal.ltisys.TransferFunctionDiscrete:
+def _pseudo_velocity_coeffs(omega, Q, T):
     """
-    Generate the transfer function
-        H(s) = L{x"(t)}(s) / L{y"(t)}(s) = X(s)/Y(s)
-    for the PDE
-        x" + (2ζω)x' + (ω²)x = (2ζω)y' + (ω²)y
+    Calculate the coefficients of the Z-domain transfer function for the
+    pseudovelocity according to ISO 18431-4
+
+    :param omega: the natural frequency of the system
+    :param Q: the quality factor of the system
+    :param T: the time step in seconds
+    :return: the coefficients of the Z-domain transfer function b, a
 
     .. seealso::
 
-        - `An Introduction To The Shock Response Spectrum, Tom Irvine, 9 July 2012 <http://www.vibrationdata.com/tutorials2/srs_intr.pdf>`_
+        - `ISO 18431-4 Mechanical vibration and shock — Signal processing — Part 4: Shock-response spectrum analysis`
+          Explicit implementations of digital filter coefficients for shock spectra.
+    """
+    A = omega*T/(2.*Q)
+    B = omega*T*np.sqrt(1. - 1./(4.*(Q**2)))
+    C = T*(omega**2)
+    q = (1./(2.*(Q**2.)) - 1.)/np.sqrt(1. - 1./(4.*(Q**2.)))
+
+    b = (
+        ((1. - np.exp(-A)*np.cos(B))/Q - q*np.exp(-A)*np.sin(B) - omega*T)/C,
+        (2.*np.exp(-A)*np.cos(B)*omega*T - (1. - np.exp(-2.*A))/Q + 2.*q*np.exp(-A)*np.sin(B))/C,
+        (-np.exp(-2.*A)*(omega*T + 1./Q) + np.exp(-A)*np.cos(B)/Q - q*np.exp(-A)*np.sin(B))/C,
+        )
+    a = (
+        1.,
+        -2.*np.exp(-A)*np.cos(B),
+        np.exp(-2.*A),
+        )
+
+    return b, a
+
+
+def pseudo_velocity(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```

Comment:
abbreviated form:
```suggestion
def pseudo_vel(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
```

non-abbreviated form: as-is 👍